pmdastatsd.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2019 Miroslav Foltýn.  All Rights Reserved.
.\" Copyright (c) 2019 Red Hat.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.TH PMDASTATSD 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmdastatsd\f1 \- \f3StatsD\f1 protocol performance metric domain agent (PMDA)
.SH SYNOPSIS
\f3$PCP_PMDAS_DIR/statsd/pmdastatsd\f1
[\f3\-Z\f1 \f2maximum udp packet size\f1]
[\f3\-P\f1 \f2port\f1]
[\f3\-v\f1]
[\f3\-g\f1]
[\f3\-o\f1 \f2debug output filename\f1]
[\f3\-s\f1]
[\f3\-r\f1 \f2parser type\f1]
[\f3\-a\f1 \f2port\f1]
[\f3\-z\f1 \f2maximum of unprocessed packets\f1]
.SH DESCRIPTION
.B StatsD 
is simple, text-based UDP protocol for receiving monitoring data of applications
in architecture client-server. 
.B pmdastatsd
is an Performance Metrics Domain Agent that collects
.B StatsD
data, aggregates them and makes them available to any Performance Co-Pilot client,
which is ideal for easily tracking stats in your application.
.PP
The
.B statsd
PMDA supports Counter, Gauge and Duration (with instances for minimum, maximum, median,
average, 90th percentile, 95th percentile, 99th percentile, count and standard deviation) metric types,
offers multiple parsing options:
.B Ragel
or
.B handwritten/custom parser
, offers multiple aggregating options for duration metric type:
.B basic histogram 
or
.B HDR histogram
, supports custom form of 
.B labels
,
.B logging
, exports multiple metrics about itself and may be configured either with ini files or command line options.
.SH CONFIGURATION
.PP
A brief description of the
.B pmdastatsd
command line options follows:
.TP 5
.B \-Z, \-\-max\-udp=<value>
Maximum allowed packet size, any larger then this will be thrown away. Default: 
.I 1472
.TP
.B \-P, \-\-port=<value>
Which port agent will listen to for incoming trafic. Default:
.I 8125
.TP
.B \-v, \-\-verbose
Verbose flag. Prints some very basic info about agent execution into logfile. Default:
.I 0
.TP
.B \-g, \-\-debug
Debug flag. Prints detailed info about agent execution. With this on, you may get quite large logfiles 
as messages regarding every packet parsing, throwing away of data and PMNS updates will be outputted. Default:
.I 0
.TP
.B \-o, \-\-debug\-output\-filename=<value>
You can send USR1 signal that 'asks' agent to output basic information about all aggregated metric
into a file with this name in agent's directory. Default:
.I 0
.TP
.B \-s, \-\-version
Flag controlling whether or not to log current agent version on start. Default:
.I 0
.TP
.B \-p, \-\-parser\-type=<value>
Flag specifying which parser to use for incoming packets, Basic =
.I 0
, Ragel =
.I 1
\[char46] Default: 
.I 0
.TP
.B \-a, \-\-duration\-aggregation\-type=<value>
Flag specifying which aggregation scheme to use for duration metrics, basic histogram = 
.I 0
, HDR histogram =
.I 1
\[char46] Default:
.I 1
.TP
.B \-z, \-max\-unprocessed\-packets=<value>
Maximum size of packet queue that the agent will save in memory. There are 2 queues: one for packets that are waiting to be parsed and one for parsed packets before they are aggregated. Default: 
.I 2048
.TP 0
Agent also looks for 
.B pmdastatsd.ini
file in 
.BR $PCP_PMDAS_DIR/statsd
directory. There, same options may be specified, albeit with slightly different names, list of which follows:
.RS 5
.P
.B max_udp_packet_size=<value>
.P
.B port=<value>
.P
.B verbose=<value>
.P
.B debug=<value>
.P
.B debug_output_filename=<value>
.P
.B version=<value>
.P
.B parser_type=<value>
.P
.B duration_aggregation_type=<value>
.P
.B max_unprocessed_packets=<value>
.RE
.P
Should an option be specified in both 
.B pmdastatsd.ini
and command line, then the latter takes precedence. Most of the time you will want to configure the agent with an ini file, as the agent should never be executed directly.
.P
Location of the log file.  By default, a log file named
.I statsd.log
is written in the current directory of
.BR pmcd (1)
when
.B pmdastatsd
is started, i.e.
.BR $PCP_LOG_DIR/pmcd .
If the log file cannot
be created or is not writable, output is written to the standard error instead.

.SH INSTALLATION
If you want to install the 
.B pmdastatsd
, do the following as root:
.PP
.ft CW
.nf
.in +0.5i
# cd $PCP_PMDAS_DIR/statsd
# ./Install
.in
.fi
.ft 1
.PP
To uninstall, do the following as root:
.PP
.ft CW
.nf
.in +0.5i
# cd $PCP_PMDAS_DIR/statsd
# ./Remove
.in
.fi
.ft 1
.PP
.B pmdastatsd
is launched by
.BR pmcd (1)
and should never be executed directly.
The Install and Remove scripts notify
.BR pmcd (1)
when the agent is installed or removed.
.SH USAGE
Once started,
.B pmdastatsd
will listen on specified
.I port
for any content in a form of:
.RS 4
.P
.B <metricname>:<value>|<type>
.RE
.P
There may be multiple such messages in single datagram, split by a newline character, so this:
.RS 4
.P
.B <metricname>:<value>|<type>\[rs]n<metricname>:<value>|<type>
.RE
.P
is valid as well.
Constraints for each of the above variables are:
.RS 4
.P
.B <metricname> = [a-z][a-zA-Z0-9_.]*
.P
.B <value>      = described further in each metric type subsection
.P
.B <type>       = One of the following: "c", "g" or "ms
.RE
.P
If 
.I verbose
logging is turned on, agent will log every message parsed and related failures.
.P
All recorded metrics will, if parsed and aggregated successfully, be made available under
.B statsd.*
namespace.
.SS 1 Counter metric
Stores metrics as simple counters, adding any incoming values to already existing ones.
.RS 4
.P
.B <metricname>:<value>|c
.RE
.P
Where 
.BI value
is positive number.
.P
.B Example:
.P
After aggregating following messages:
.RS 4
.P
.B metric:20|c

.B metric:10|c

.B metric:3.3|c
.RE
.P
Value available to PCP will be:
.RS 4
.P

.B pminfo \-f statsd.metric

.B \-> inst [0 or '/'] value 33.3
.RE
.SS 2 Gauge metric
.P
Stores metrics as modifiable values, with an option to either set, increment or decrement values.
.RS 4
.P
.B <metricname>:<value>|g
.RE
.P
Where
.BI value
can be in a form of:
.RS 4
.P
.B '\-{value}'
, when negative value is supplied agent will substract value stored with the value passed

.B '+{value}'
, when positive value with a leading plus sign is supplied, agent will add passed value to the value stored

.B '{value}'
, when value without any leading sign is supplied, agent will set the metric to the passed value
.RE
.P
Initial value for metric of gauge type is 0.
.P
.B Example:
.P
After aggregating following messages:
.RS 4
.P
.B metric:20|g

.B metric:+10|g

.B metric:-3.3|g
.RE
.P
Value available to PCP will be:
.RS 4
.P
.B pminfo \-f statsd.metric

.B \-> inst [0 or '/'] value 26.7
.RE
.SS 3 Duration metric
.P
Aggregates values either via HDR histogram or simply stores all values and then calculates instances from all values received.
.RS 4
.P
.B <metricname>:<value>|ms
.RE
.P
Where
.BI value
is a positive number.
.P
.B Example:
.P
With larger message count, the values may vary based on selected duration aggregation scheme.
.P
After aggregating following messages:
.RS 4
.P
.B metric:10|ms

.B metric:20|ms
.RE
.P
Values available to PCP will be:
.RS 4
.P
.B pminfo \-f statsd.metric

.B \-> inst[0 or '/min'] value 10

.B \-> inst[1 or '/max'] value 20

.B \-> inst[2 or '/median'] value 10

.B \-> inst[3 or '/average'] value 15

.B \-> inst[4 or '/percentile90'] value 20

.B \-> inst[5 or '/percentile95'] value 20

.B \-> inst[6 or '/percentile99'] value 20

.B \-> inst[7 or '/count'] value 2

.B \-> inst[8 or '/std_deviation'] value 5
.RE
.P
.B Note:
.P
Once you send given 
.I metricname
with specified
.I type
, agent will no longer aggregate any messages with same.
.I metricname
but different
.I type
and will throw them away.
.SS 4 Labels
StatsD datagrams may also contain key:value pairs separated by commas like so:
.RS 4
.P
.B metric,tagX=X,tagW=W:5|c
.P
OR
.P
.B metric:5|ms|#tagX:X,tagW:W
.RE
.P
Where 
.BI tagX
is a 
.I key
,
.BI X
is a 
.I value
and
.BI tagW
is a 
.I key
,
.BI W
is a 
.I value
\[char46]
.P
Both 
.I key 
and 
.I value
of such pair are 
.BI [a\-ZA\-Z0\-9_.]{1,}
\[char46]
.P
Both formats are interchangeable and you may combine them together. When 
.I key
is not unique, right most
.I value
takes precendence. This is valid:
.RS 4
.P
.B metric,tagX=1:5|c|#tagX:2
.RE
.P
Pair with 
.I key
.I tagX
will have value of 2.
.P
You may use these labels to map specific values to some PCP instances. PCP labels are also assigned to these PCP instances. Pairs are ordered by
.I key
in resulting instance name and label descriptor.
.P
Single label:
.RS 4
.P
.B metric,tagX=X:5|c
.RE
.P
Such payload would map to PCP as follows (non-related labels were ommited):
.RS 4
.P
.B pminfo \-f \-\-labels statsd.metric

.B \->

.B inst [0 or '/tagX=X'] value 5 

.B inst [0 or '/tagX=X'] labels {'tagX':'X'}
.RE
.P
As shown earlier you may also send payload with multiple labels. When multiple labels are supplied they are split in instance name by '::'. Example:
.RS 4
.P
.B metric,tagX=X,tagW=W:5|c
.RE
.P
This resolves to:
.RS 4
.P
.B pminfo \-f \-\-labels statsd.metric

.B \->

.B inst [0 or '/tagX=X::tagW=W'] value 5 

.B inst [0 or '/tagX=X::tagW=W'] labels {'tagX':'X','tagW':'W'}
.RE
.P
.B Note:
.P
Be mindful of the fact that duration metric type already maps to instances even without any labels. Sending labeled value to a such metric creates another 9 (as there are that many hardcoded) instances.
.P
Example:
.RS 4
.P
.B metric:200|ms

.B metric:100|ms

.B metric,target=cpu2:10|ms

.B metric,target=cpu2:100|ms

.B metric,target=cpu2:1000|ms
.RE
.P
Creates 18 instances. Duration data type and label name compose instance name in following manner:
.RS 4
.P
.B pminfo \-f \-\-labels statsd.metric

.B ...

.B inst [10 or '/max::target=cpu0'] value 1000

.B inst [10 or '/max::target=cpu0'] labels {'target':'cpu0'}

.B ...
.RE
.P
.SS 5 Hardcoded stats
Agent also exports metrics about itself:
.TP 5
.B statsd.pmda.received
Number of datagrams that the agent has received
.TP
.B statsd.pmda.parsed
Number of datagrams that were successfully parsed
.TP
.B statsd.pmda.dropped
Number of datagrams that were dropped
.TP
.B statsd.pmda.aggregated
Number of datagrams that were aggregated
.TP
.B statsd.pmda.metrics_tracked
This metric has 3 instances.
.B counter
- Number of tracked counter metrics.
.B gauge
- Number of tracked gauge metrics.
.B duration
- Number of tracked duration metrics.
.B total
- Number of tracked metrics total.
.TP
.B statsd.pmda.settings.max_udp_packet_size
Maximum UDP packet size
.TP
.B statsd.pmda.settings.max_unprocessed_packets
Maximum size of unprocessed packets Q
.TP
.B statsd.pmda.settings.verbose
Verbosity flag
.TP
.B statsd.pmda.settings.debug
Debug flag
.TP
.B statsd.pmda.settings.debug_output_filename
Debug output filename
.TP
.B statsd.pmda.settings.port
Port that is listened to
.TP
.B statsd.pmda.settings.parser_type
Used parser type
.TP
.B statsd.pmda.settings.duration_aggregation_type
Used duration aggregation type
.P
These names are blacklisted for user usage. No messages with these names will processed. While not yet reserved, whole 
.B statsd.pmda.*
namespace is not recommended to use for user metrics.

.SH FILES
.PD 0
.TP 10
.B $PCP_PMCDCONF_PATH
command line options used to launch
.B pmdastatsd
.TP 10
.B $PCP_PMDAS_DIR/statsd/Install
installation script for the
.B pmdastatsd
agent
.TP 10
.B $PCP_PMDAS_DIR/statsd/Remove
undo installation script for the 
.B pmdastatsd
agent
.TP 10
.B $PCP_LOG_DIR/pmcd/statsd.log
default log file for error messages and other information from
.B pmdastatsd
.PD
.SH "PCP ENVIRONMENT"
Environment variables with the prefix
.B PCP_
are used to parameterize the file and directory names
used by PCP.
On each installation, the file
.I /etc/pcp.conf
contains the local values for these variables.
The
.B $PCP_CONF
variable may be used to specify an alternative
configuration file,
as described in
.BR pcp.conf (5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmcd (1),
.BR pcp.conf (5)
and
.BR pcp.env (5).


src/pmdastatsd.c
#ifndef _TEST_TARGET

#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#include <chan/chan.h>
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <signal.h>
#include <pcp/pmapi.h>

#include "pmdastatsd.h"
#include "config-reader.h"
#include "network-listener.h"
#include "aggregators.h"
#include "aggregator-metrics.h"
#include "aggregator-stats.h"
#include "pmda-callbacks.h"
#include "dict-callbacks.h"
#include "utils.h"
#include "../domain.h"

#define VERSION 0.9

void output_request_handler(int num) {
    if (num == SIGUSR1) {
        DEBUG_LOG("Handling SIGUSR1.");
        aggregator_request_output();
    }
    if (num == SIGINT) {
        DEBUG_LOG("Handling SIGINT.");
        set_exit_flag();
    }
}

#define SET_INST_NAME(instance, name, index) \
    instance[index].i_inst = index; \
    len = pmsprintf(buff, 20, "%s", name) + 1; \
    instance[index].i_name = (char*) malloc(sizeof(char) * len); \
    ALLOC_CHECK("Unable to allocate memory for static PMDA instance descriptor."); \
    memcpy(instance[index].i_name, buff, len);

/**
 * Registers hardcoded instances before PMDA initializes itself fully
 * @arg pmda - PMDA extension structure (contains agent-specific private data)
 */
static void
create_statsd_hardcoded_instances(struct pmda_data_extension* data) {
    size_t len = 0;
    char buff[20];
    size_t hardcoded_count = 3;

    data->pcp_instance_domains = (pmdaIndom*) malloc(hardcoded_count * sizeof(pmdaIndom));
    ALLOC_CHECK("Unable to allocate memory for static PMDA instance domains.");
    
    pmdaInstid* stats_metric_counters_indom = (pmdaInstid*) malloc(sizeof(pmdaInstid) * 4);
    ALLOC_CHECK("Unable to allocate memory for static PMDA instance domain descriptor.");
    data->pcp_instance_domains[0].it_indom = STATS_METRIC_COUNTERS_INDOM;
    data->pcp_instance_domains[0].it_numinst = 4;
    data->pcp_instance_domains[0].it_set = stats_metric_counters_indom;
    SET_INST_NAME(stats_metric_counters_indom, "counter", 0);
    SET_INST_NAME(stats_metric_counters_indom, "gauge", 1);
    SET_INST_NAME(stats_metric_counters_indom, "duration", 2);
    SET_INST_NAME(stats_metric_counters_indom, "total", 3);

    pmdaInstid* statsd_metric_default_duration_indom = (pmdaInstid*) malloc(sizeof(pmdaInstid) * 9);
    ALLOC_CHECK("Unable to allocate memory for static PMDA instance domain descriptors.");
    data->pcp_instance_domains[1].it_indom = STATSD_METRIC_DEFAULT_DURATION_INDOM;
    data->pcp_instance_domains[1].it_numinst = 9;
    data->pcp_instance_domains[1].it_set = statsd_metric_default_duration_indom;
    SET_INST_NAME(statsd_metric_default_duration_indom, "/min", 0);
    SET_INST_NAME(statsd_metric_default_duration_indom, "/max", 1);
    SET_INST_NAME(statsd_metric_default_duration_indom, "/median", 2);
    SET_INST_NAME(statsd_metric_default_duration_indom, "/average", 3);
    SET_INST_NAME(statsd_metric_default_duration_indom, "/percentile90", 4);
    SET_INST_NAME(statsd_metric_default_duration_indom, "/percentile95", 5);
    SET_INST_NAME(statsd_metric_default_duration_indom, "/percentile99", 6);
    SET_INST_NAME(statsd_metric_default_duration_indom, "/count", 7);
    SET_INST_NAME(statsd_metric_default_duration_indom, "/std_deviation", 8);

    pmdaInstid* statsd_metric_default_indom = (pmdaInstid*) malloc(sizeof(pmdaInstid));
    ALLOC_CHECK("Unable to allocate memory for default dynamic metric instance domain descriptior");
    data->pcp_instance_domains[2].it_indom = STATSD_METRIC_DEFAULT_INDOM;
    data->pcp_instance_domains[2].it_numinst = 1;
    data->pcp_instance_domains[2].it_set = statsd_metric_default_indom;
    SET_INST_NAME(statsd_metric_default_indom, "/", 0);

    data->pcp_instance_domain_count = hardcoded_count;
    data->pcp_hardcoded_instance_domain_count = hardcoded_count;
}

/**
 * Registers hardcoded metrics before PMDA initializes itself fully
 * @arg pmda - PMDA extension structure (contains agent-specific private data)
 */
static void
create_statsd_hardcoded_metrics(struct pmda_data_extension* data) {
    size_t i;
    size_t hardcoded_count = 15;
    data->pcp_metrics = (pmdaMetric*) malloc(hardcoded_count * sizeof(pmdaMetric));
    ALLOC_CHECK("Unable to allocate space for static PMDA metrics.");
    // helper containing only reference to priv data same for all hardcoded metrics
    static struct pmda_metric_helper helper;
    size_t agent_stat_count = 7;
    helper.data = data;
    for (i = 0; i < hardcoded_count; i++) {
        data->pcp_metrics[i].m_user = &helper;
        data->pcp_metrics[i].m_desc.pmid = pmID_build(STATSD, 0, i);
        data->pcp_metrics[i].m_desc.sem = PM_SEM_INSTANT;
        if (i < agent_stat_count) {
            data->pcp_metrics[i].m_desc.type = PM_TYPE_U64;
            if (i == 4) {
                data->pcp_metrics[i].m_desc.indom = STATS_METRIC_COUNTERS_INDOM;
            } else {
                data->pcp_metrics[i].m_desc.indom = PM_INDOM_NULL;
            }            
        } else {
            if (i == 7) {
                data->pcp_metrics[i].m_desc.type = PM_TYPE_U64;
            } else if (i < 11 || i == 12) {
                data->pcp_metrics[i].m_desc.type = PM_TYPE_U32;
            } else {
                data->pcp_metrics[i].m_desc.type = PM_TYPE_STRING;
            }
            data->pcp_metrics[i].m_desc.indom = PM_INDOM_NULL;
        }
        if (i == 5 || i == 6) {
            // time_spent_parsing / time_spent_aggregating
            data->pcp_metrics[i].m_desc.units.dimSpace = 0;
            data->pcp_metrics[i].m_desc.units.dimTime = 0;
            data->pcp_metrics[i].m_desc.units.dimCount = 0;
            data->pcp_metrics[i].m_desc.units.pad = 0;
            data->pcp_metrics[i].m_desc.units.scaleSpace = 0;
            data->pcp_metrics[i].m_desc.units.scaleTime = PM_TIME_NSEC;
            data->pcp_metrics[i].m_desc.units.scaleCount = 1;
        } else {
            // rest
            memset(&data->pcp_metrics[i].m_desc.units, 0, sizeof(pmUnits));
        }
    }
    data->pcp_metric_count = hardcoded_count;
    data->pcp_hardcoded_metric_count = hardcoded_count;
}

static void
free_shared_data(struct agent_config* config, struct pmda_data_extension* data) {
    // frees config
    free(config->debug_output_filename);
    // remove metrics dictionary and related
    dictRelease(data->metrics_storage->metrics);
    // privdata will be left behind, need to remove manually
    free(data->metrics_storage->metrics_privdata);
    pthread_mutex_destroy(&data->metrics_storage->mutex);
    free(data->metrics_storage);
    // remove stats dictionary and related
    free(data->stats_storage->stats->metrics_recorded);
    free(data->stats_storage->stats);
    pthread_mutex_destroy(&data->stats_storage->mutex);
    free(data->stats_storage);
    // free instance map
    dictRelease(data->instance_map);
    // clear PCP metric table
    size_t i;
    for (i = 0; i < data->pcp_metric_count; i++) {
        size_t j = data->pcp_hardcoded_metric_count;
        if (i == j) {
            free(data->pcp_metrics[i].m_user);
        }
        if (!(i < j)) {
            free(data->pcp_metrics[i].m_user);
        }
    }
    free(data->pcp_metrics);
    // clear PCP instance domains
    for (i = 0; i < data->pcp_instance_domain_count; i++) {
        int j = 0;
        for (j; j < data->pcp_instance_domains[i].it_numinst; j++) {
            free(data->pcp_instance_domains[i].it_set[j].i_name);
        }
        free(data->pcp_instance_domains[i].it_set);
    }
    free(data->pcp_instance_domains);
    // Release PMNS tree
    pmdaTreeRelease(data->pcp_pmns);
}

/**
 * Initializes structure which is used as private data container accross all PCP related callbacks
 * @arg args - All args passed to 'PCP exchange' thread
 */
static void
init_data_ext(
    struct pmda_data_extension* data,
    struct agent_config* config,
    struct pmda_metrics_container* metrics_storage,
    struct pmda_stats_container* stats_storage
) {
    static dictType instance_map_callbacks = {
        .hashFunction	= str_hash_callback,
        .keyCompare		= str_compare_callback,
        .keyDup		    = str_duplicate_callback,
        .keyDestructor	= str_hash_free_callback,
    };
    data->config = config;
    create_statsd_hardcoded_metrics(data);
    create_statsd_hardcoded_instances(data);
    data->metrics_storage = metrics_storage;
    data->stats_storage = stats_storage;
    data->instance_map = dictCreate(&instance_map_callbacks, NULL);
    data->generation = -1; // trigger first mapping of metrics for PMNS 
    data->notify = 0;
}

static void
main_PDU_loop(pmdaInterface* dispatch) {
    for(;;) {
        if (check_exit_flag() != 0) break;
        if (__pmdaMainPDU(dispatch) < 0) break;
    }
}

int
main(int argc, char** argv)
{
    struct sigaction new_action, old_action;

    /* Set up the structure to specify the new action. */
    new_action.sa_handler = output_request_handler;
    sigemptyset(&new_action.sa_mask);
    new_action.sa_flags = SA_INTERRUPT;

    sigaction (SIGUSR1, NULL, &old_action);
    if (old_action.sa_handler != SIG_IGN) {
        sigaction (SIGUSR1, &new_action, NULL);
    }
    sigaction (SIGINT, NULL, &old_action);
    if (old_action.sa_handler != SIG_IGN) {
        sigaction (SIGINT, &new_action, NULL);
    }

    struct agent_config config = { 0 };
    struct pmda_data_extension data = { 0 };
    pthread_t network_listener;
    pthread_t parser;
    pthread_t aggregator;
    pmdaInterface dispatch = { 0 };

    int sep = pmPathSeparator();
    char config_file_path[MAXPATHLEN];
    char help_file_path[MAXPATHLEN];
    pmsprintf(
        config_file_path,
        MAXPATHLEN,
        "%s" "%c" "statsd" "%c" "pmdastatsd.ini",
        pmGetConfig("PCP_PMDAS_DIR"),
        sep, sep);
    pmsprintf(
        help_file_path,
        MAXPATHLEN,
        "%s%c" "statsd" "%c" "help",
        pmGetConfig("PCP_PMDAS_DIR"),
        sep, sep);

    pmSetProgname(argv[0]);
    pmdaDaemon(&dispatch, PMDA_INTERFACE_7, pmGetProgname(), STATSD, "statsd.log", help_file_path);

    read_agent_config(&config, &dispatch, config_file_path, argc, argv);
    init_loggers(&config);
    pmdaOpenLog(&dispatch);
    if (config.debug) {
        print_agent_config(&config);
    }
    if (config.show_version) {
        pmNotifyErr(LOG_INFO, "Version: %f", VERSION);
    }

    struct pmda_metrics_container* metrics = init_pmda_metrics(&config);
    struct pmda_stats_container* stats = init_pmda_stats(&config);
    init_data_ext(&data, &config, metrics, stats);

    chan_t* network_listener_to_parser = chan_init(config.max_unprocessed_packets);
    if (network_listener_to_parser == NULL) {
        DIE("Unable to create channel network listener -> parser.");        
    }
    ALLOC_CHECK("Unable to create channel network listener -> parser.");
    chan_t* parser_to_aggregator = chan_init(config.max_unprocessed_packets);
    if (parser_to_aggregator == NULL) {
        DIE("Unable to create channel parser -> aggregator.");        
    }
    ALLOC_CHECK("Unable to create channel parser -> aggregator.");
    struct network_listener_args* listener_args = create_listener_args(&config, network_listener_to_parser);
    struct parser_args* parser_args = create_parser_args(&config, network_listener_to_parser, parser_to_aggregator);
    struct aggregator_args* aggregator_args = create_aggregator_args(&config, parser_to_aggregator, metrics, stats);

    int pthread_errno = 0; 
    pthread_errno = pthread_create(&network_listener, NULL, network_listener_exec, listener_args);
    PTHREAD_CHECK(pthread_errno);
    pthread_errno = pthread_create(&parser, NULL, parser_exec, parser_args);
    PTHREAD_CHECK(pthread_errno);
    pthread_errno = pthread_create(&aggregator, NULL, aggregator_exec, aggregator_args);
    PTHREAD_CHECK(pthread_errno);

    pmSetProcessIdentity(config.username);

    if (dispatch.status != 0) {
        pthread_exit(NULL);
    }
    dispatch.version.seven.fetch = statsd_fetch;
	dispatch.version.seven.desc = statsd_desc;
	dispatch.version.seven.text = statsd_text;
	dispatch.version.seven.instance = statsd_instance;
	dispatch.version.seven.pmid = statsd_pmid;
	dispatch.version.seven.name = statsd_name;
	dispatch.version.seven.children = statsd_children;
	dispatch.version.seven.label = statsd_label;
    // Callbacks
	pmdaSetFetchCallBack(&dispatch, statsd_fetch_callback);
    pmdaSetLabelCallBack(&dispatch, statsd_label_callback);
    pmdaSetData(&dispatch, (void*) &data);
    pmdaSetFlags(&dispatch, PMDA_EXT_FLAG_HASHED);
    pmdaInit(
        &dispatch,
        data.pcp_instance_domains,
        data.pcp_instance_domain_count,
        data.pcp_metrics,
        data.pcp_metric_count
    );
    pmdaConnect(&dispatch);
    main_PDU_loop(&dispatch);
    
    if (pthread_join(network_listener, NULL) != 0) {
        DIE("Error joining network network listener thread.");
    } else {
        VERBOSE_LOG("Network listener thread joined.");
        set_parser_exit();
    }
    if (pthread_join(parser, NULL) != 0) {
        DIE("Error joining datagram parser thread.");
    } else {
        VERBOSE_LOG("Parser thread joined.");
        set_aggregator_exit();
    }
    if (pthread_join(aggregator, NULL) != 0) {
        DIE("Error joining datagram aggregator thread.");
    } else {
        VERBOSE_LOG("Aggregator thread joined.");
    }

    free_shared_data(&config, &data);
    free(listener_args);
    free(parser_args);
    free(aggregator_args);

    chan_close(network_listener_to_parser);
    chan_close(parser_to_aggregator);
    chan_dispose(network_listener_to_parser);
    chan_dispose(parser_to_aggregator);
    return EXIT_SUCCESS;
}

#endif
